File: GameEngine_IOS/GameEngine/AudioEngine/EntityPlayer.swift
import Foundation


/// Is going to demand the reproduction of audio due to the control of the entities
open class EntityPlayer : GenericPlayer {
    
    
    /// If the entity is far away from the player more that the threshold is not going to be listened
    fileprivate let SOUND_THRESHOLD : Float = 50.0
    
    /// First if going to make a rough approximation only after is going to check if exists a crash
    fileprivate let ROUGH_THRESHOLD : Float = 10.0
    fileprivate let CRASH_THRESHOLD : Float = 2.0
    
    
    /// List of sound sources available to the entity player
    fileprivate var sourcesAvailable : [AudioSource]
    
    
    /// The player can always play audio
    fileprivate var playerSource : AudioSource!
    
    
    /// Sources assigned to the entities
    fileprivate var sourcesAssign : [Entity : AudioSource]
    
    /// Initializer of the entity player
    ///
    /// - Parameter sourcesAvailable: The list of sources available
    public init(sourcesAvailable : [AudioSource]) {
        self.sourcesAvailable = sourcesAvailable;
        if ((!self.sourcesAvailable.isEmpty)) {
            playerSource = self.sourcesAvailable.removeLast()
        } else {
            playerSource = nil
        }
        self.sourcesAssign = [Entity : AudioSource]();
    }
    
    
    /// Set the listener position and reproduces the sound of the player
    ///
    /// - Parameters:
    ///   - library: Set of sounds that can use to provide sound feedback to the user
    ///   - player: The player of the scene and also the listener
    fileprivate func playPlayer(library : [TAudioEnum: AudioBuffer], player : Player) {
        self.positioningListener(position: player.position);
        
        // If the player is running plays the sound of the footsteps
        if (player.getCurrentSpeed() == 0) {
            if (self.isPlaying(source: playerSource)) {
                self.pause(source: playerSource);
            }
        } else {
            let pitch = player.getCurrentSpeed() / 15.0
            if (self.isPaused(source: playerSource)) {
                self.continuePlaying(source: playerSource);
            } else {
                if (!self.isPlaying(source: playerSource)) {
                    let footStepsBuffer : AudioBuffer = library[TAudioEnum.footsteps]!;
                    self.setPitch(source: playerSource, pitch: pitch);
                    self.setVolume(source: playerSource, volume: 1.0);
                    self.play(source: playerSource, buffer: footStepsBuffer);
                    self.setLoop(source: self.playerSource, loop: true);
                }
            }
            self.setPosition(source: playerSource, position: player.position);
        }
    }
    
    /// Makes one aproximation component by compoenent (Faster that usual distance)
    ///
    /// - Parameters:
    ///   - position1: The position one to take in account
    ///   - position2: The position two entity to take in account
    ///   - threshold: The threshold that is going to use to the match
    /// - Returns:  False = Not passed in the rough approximation
    ///             True  = Passed in the rough approximation
    fileprivate func roughApproximation(position1 : Vector3f, position2 : Vector3f, threshold : Float) -> Bool {
        return (abs(position1.x - position2.x) < threshold) &&
            (abs(position1.y - position2.y) < threshold) &&
            (abs(position1.z - position2.z) < threshold);
    }
    
    
    /// Computes the distance between two points
    ///
    /// - Parameters:
    ///   - position1: The position one to take in account
    ///   - position2: The position two entity to take in account
    /// - Returns: The square of the distance between two positions
    fileprivate func squareDistance(position1 : Vector3f, position2 : Vector3f) -> Float {
        let xDiff : Float = abs(position1.x - position2.x);
        let yDiff : Float = abs(position1.y - position2.y);
        let zDiff : Float = abs(position1.z - position2.z);
        
        return (xDiff * xDiff) + (yDiff * yDiff) + (zDiff * zDiff);
    }
    
    
    /// Indicates if an entity passes the threshold to be considered
    ///
    /// - Parameters:
    ///   - pPosition: The position of player to take in account
    ///   - ePosition: The position of entity to take in account
    /// - Returns: False the player can not listen the entity
    ///             True the entity can be listen by the player
    fileprivate func passesThreshold(pPosition : Vector3f, ePosition : Vector3f) -> Bool {
        return roughApproximation(position1: pPosition, position2: ePosition, threshold: SOUND_THRESHOLD);
    }
    

    /// Assign audio sources to the entities
    ///
    /// - Parameters:
    ///   - player: The player of the scene and also the listener
    ///   - entities: Set of entities that should provide sound feedback
    fileprivate func assignSources(player : Player, entities : [Entity]) {
        let pPosition : Vector3f = player.position;
        for entity in entities {
            if (passesThreshold(pPosition: pPosition, ePosition: entity.position)) {
                // Can not assign nothing
                if (sourcesAvailable.isEmpty) {
                    //Is going to continue searching maybe there is anothers that do not pass the threshold
                    continue;
                }
                // Check is a source was already assign
                if (sourcesAssign[entity] == nil) {
                    let aSource : AudioSource = sourcesAvailable.removeLast();
                    sourcesAssign[entity] = aSource;
                }
            } else {
                // Not passed threshold
                if (sourcesAssign[entity] != nil) {
                    let aSource = sourcesAssign.removeValue(forKey: entity)!;
                    self.stop(source: aSource);
                    self.sourcesAvailable.append(aSource);
                }
            }
        }
    }
    

    /// Check is an entity crashed against the player
    ///
    /// - Parameters:
    ///   - library: Set of sounds that can use to provide sound feedback to the user
    ///   - entity: The entity to check
    ///   - player: The position of the player that is going to check the crash
    /// - Returns:  False = No crash was detected
    ///             True = A crash was detected and the sound was played
    fileprivate func playCrash(library : [TAudioEnum : AudioBuffer], entity : Entity, player : Vector3f) -> Bool {
        if (roughApproximation(position1: player, position2: entity.position, threshold: ROUGH_THRESHOLD)
            && (squareDistance(position1: player, position2: entity.position) < CRASH_THRESHOLD)) {
            let source : AudioSource = sourcesAssign[entity]!;
            let tEntity : TEntity = entity.genericEntity.objectType;
            
            switch(tEntity) {
            case .banana_tree:
                break;
            case .fern:
                break;
            case .flower:
                break;
            case .grass:
                break;
            case .marble:
                if (!isPlaying(source: source) && (source.getBuffer() == nil)) {
                    let breakingWood : AudioBuffer = library[TAudioEnum.breakingWood]!;
                    self.setPitch(source: source, pitch: 1.0);
                    self.setVolume(source: source, volume: 1.0);
                    self.setPosition(source: source, position: entity.position);
                    self.play(source: source, buffer: breakingWood);
                    self.setLoop(source: source, loop: false);
                }
                break;
            case .player:
                break;
            case .tree:
                let breakingWood : AudioBuffer! = library[TAudioEnum.breakingWood];
                if (!isPlaying(source: source) || ((source.getBuffer() == nil) || (source.getBuffer() != breakingWood))) {
                    self.setPitch(source: source, pitch: 1.0);
                    self.setVolume(source: source, volume: 1.0);
                    self.setPosition(source: source, position: entity.position);
                    self.play(source: source, buffer: breakingWood);
                    self.setLoop(source: source, loop: false);
                }
                break;
            }
            
            return true;
        } else {
            return false;
        }
    }
    

    /// Reproduces the sound of the entities
    ///
    /// - Parameters:
    ///   - library: Set of sounds that can use to provide sound feedback to the user
    ///   - entities: Set of entities that should provide sound feedback
    ///   - player: The player of the scene and also the listener
    fileprivate func playEntities(library : [TAudioEnum : AudioBuffer], entities : [Entity]!, player : Player) {
        if (entities != nil) {
            let pPosition : Vector3f = player.position;
            
            for entity in entities {
                // Check is has one source
                let source : AudioSource! = sourcesAssign[entity];
                if (source == nil) {
                    continue;
                }
                
                if (!playCrash(library: library, entity: entity, player: pPosition)) {
                    // If not played a crash
                    if (entity.genericEntity.objectType != TEntity.tree) {
                        continue;
                    }
                    
                    // Make birds to play in each tree
                    let breakingWood : AudioBuffer = library[TAudioEnum.breakingWood]!;
                    if ((source.getBuffer() != nil) && (source.getBuffer() == breakingWood)) {
                        continue;
                    }
                    if (!isPlaying(source: source)) {
                        let falconBuffer : AudioBuffer = library[TAudioEnum.falcon]!;
                        self.setPitch(source: source, pitch: 1.0);
                        self.setVolume(source: source, volume: 1.0);
                        self.setPosition(source: source, position: entity.position);
                        self.play(source: source, buffer: falconBuffer);
                        self.setLoop(source: source, loop: true);
                    }
                }
                
            }
        }
    }
    

    /// Plays all the sounds related with entities
    ///
    /// - Parameters:
    ///   - library: Set of sounds that can use to provide sound feedback to the user
    ///   - entities: Set of entities that should provide sound feedback
    ///   - player: The player of the scene and also the listener
    open func play(library : [TAudioEnum: AudioBuffer] , entities : [Entity], player : Player) {
        self.playPlayer(library: library, player: player);
        self.assignSources(player: player, entities: entities);
        let entitiesPlaying : [Entity] = Array(self.sourcesAssign.keys);
        
        self.playEntities(library: library, entities: entitiesPlaying, player: player);
        
    }
    

    /// Clean up because we need to clean up when we finish the program
    open func cleanUp() {
    }
}

File: GameEngine_PC/GameEngineOpenGL3_2/src/main/java/com/dferreira/gameEngine/audioEngine/EntityPlayer.java
package com.dferreira.gameEngine.audioEngine;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.dferreira.commons.Vector3f;
import com.dferreira.commons.generic_player.IAudioDescription;
import com.dferreira.commons.generic_player.IAudioSource;
import com.dferreira.commons.generic_player.IListener;
import com.dferreira.commons.generic_resources.AudioEnum;
import com.dferreira.commons.generic_resources.ModelEnum;
import com.dferreira.commons.utils.Utils;
import com.dferreira.gameEngine.models.Player;
import com.dferreira.gameEngine.models.complexEntities.Entity;

/**
 * Is going to demand the reproduction of audio due to the control of the
 * entities
 */
public class EntityPlayer {

	/**
	 * If the entity is far away from the player more that the threshold is not
	 * going to be listened
	 */
	private float SOUND_THRESHOLD = 50f;

	/**
	 * First if going to make a rough approximation only after is going to check
	 * if exists a crash
	 */
	private float ROUGH_THRESHOLD = 10;
	private float CRASH_THRESHOLD = 2;

	/**
	 * List of sound sources available to the entity player
	 * 
	 */
	private List<IAudioSource> sourcesAvailable;

	/**
	 * The player can always play audio;
	 */
	private IAudioSource playerSource;

	/**
	 * Sources assign
	 */
	private Map<Entity, IAudioSource> sourcesAssign;

	/**
	 * Constructor of the entity player
	 */
	public EntityPlayer(List<IAudioSource> sourcesAvailable) {
		this.sourcesAvailable = sourcesAvailable;
		if (!Utils.isEmpty(this.sourcesAvailable)) {
			this.playerSource = this.sourcesAvailable.remove(this.sourcesAvailable.size() - 1);
		}
		this.sourcesAssign = new HashMap<Entity, IAudioSource>();
	}

	/**
	 * Set the listener position and reproduces the sound of the player
	 * 
	 * @param library
	 *            Set of sounds that can use to provide sound feedback to the
	 *            user
	 * @param listener
	 *            Listener to put in a certain position
	 * @param player
	 *            The player of the scene and also the listener
	 */
	private void playPlayer(HashMap<AudioEnum, IAudioDescription> library, IListener listener, Player player) {
		listener.positioning(player.getPosition());

		// If the player is running plays the sound of the footsteps
		if (player.getCurrentSpeed() == 0) {
			if (playerSource.isPlaying()) {
				playerSource.pause();
			}
		} else {
			float pitch = player.getCurrentSpeed() / 30.0f;
			if (playerSource.isPaused()) {
				playerSource.setPitch(pitch);
				playerSource.continuePlaying();
			} else {
				if (!playerSource.isPlaying()) {
					IAudioDescription footStepsBuffer = library.get(AudioEnum.footsteps);
					playerSource.setPitch(pitch);
					playerSource.setVolume(1.0f);
					playerSource.play(footStepsBuffer);
					playerSource.setLoop(true);
				}
			}
			playerSource.setPosition(player.getPosition());
		}
	}

	/**
	 * 
	 * @param position1
	 *            The position one to take in account
	 * @param position2
	 *            The position two entity to take in account
	 * @param threshold
	 *            The threshold that is going to use to the match
	 * 
	 * @return False = Not passed in the rough approximation True = Passed in
	 *         the rough approximation
	 */
	private boolean roughApproximation(Vector3f position1, Vector3f position2, float threshold) {
		return (Math.abs(position1.x - position2.x) < threshold) && (Math.abs(position1.y - position2.y) < threshold)
				&& (Math.abs(position1.z - position2.z) < threshold);
	}

	/**
	 * 
	 * @param position1
	 *            The position one to take in account
	 * @param position2
	 *            The position two entity to take in account
	 * 
	 * 
	 * @return The square of the distance between two positions
	 */
	private float squareDistance(Vector3f position1, Vector3f position2) {
		float xDiff = Math.abs(position1.x - position2.x);
		float yDiff = Math.abs(position1.y - position2.y);
		float zDiff = Math.abs(position1.z - position2.z);

		return (xDiff * xDiff) + (yDiff * yDiff) + (zDiff * zDiff);
	}

	/**
	 * 
	 * @param pPosition
	 *            The position of player to take in account
	 * @param ePosition
	 *            The position of entity to take in account
	 * 
	 * @return False the player can not listen the entity True the entity can be
	 *         listen by the player
	 */
	private boolean passesThreshold(Vector3f pPosition, Vector3f ePosition) {
		return roughApproximation(pPosition, ePosition, SOUND_THRESHOLD);
	}

	/**
	 * Assign audio sources to the entities	taking in account the position of the player
	 * 
	 * @param player
	 *            The player of the scene and also the listener
	 * @param entities
	 *            Set of entities that should provide sound feedback
	 */
	private void assignSources(Player player, Entity[] entities) {
		if ((player != null) && (!Utils.isEmpty(entities))) {
			Vector3f pPosition = player.getPosition();
			for (Entity entity : entities) {
				if (passesThreshold(pPosition, entity.getPosition())) {
					// Can not assign nothing
					if (sourcesAvailable.isEmpty()) {
						continue;
					}
					// Check is a source was already assign
					if (!sourcesAssign.containsKey(entity)) {
						IAudioSource aSource = sourcesAvailable.remove(sourcesAvailable.size() - 1);
						sourcesAssign.put(entity, aSource);
					}
				} else {
					// Not passed threshold
					if (sourcesAssign.containsKey(entity)) {
						IAudioSource aSource = sourcesAssign.remove(entity);
						aSource.stop();
						this.sourcesAvailable.add(aSource);
					}
				}
			}
		}
	}

	/**
	 * Check is an entity crashed against the player
	 * 
	 * @param library
	 *            Set of sounds that can use to provide sound feedback to the
	 *            user
	 * @param entity
	 *            The entity to check
	 * @param player
	 *            The position of the player that is going to check the crash
	 * 
	 * @return False = No crash was detected True = A crash was detected and the
	 *         sound was played
	 */
	private boolean playCrash(HashMap<AudioEnum, IAudioDescription> library, Entity entity, Vector3f player) {
		if (roughApproximation(player, entity.getPosition(), ROUGH_THRESHOLD)
				&& (squareDistance(player, entity.getPosition()) < CRASH_THRESHOLD)) {
			IAudioSource source = sourcesAssign.get(entity);
			ModelEnum tEntity = entity.getGenericEntity().getObjectType();
			switch (tEntity) {
			case banana_tree:
				break;
			case fern:
				break;
			case flower:
				break;
			case grass:
				break;
			case marble:
				if (!source.isPlaying() && (source.getAudioDescription() == null)) {
					IAudioDescription breakingWood = library.get(AudioEnum.breakingWood);
					source.setPitch(1.0f);
					source.setVolume(1.0f);
					source.setPosition(entity.getPosition());
					source.play(breakingWood);
					source.setLoop(false);
				}
				break;
			case player:
				break;
			case tree:
				IAudioDescription breakingWood = library.get(AudioEnum.breakingWood);
				if (!source.isPlaying()
						|| ((source.getAudioDescription() == null) || (source.getAudioDescription() != breakingWood))) {
					source.setPitch(1.0f);
					source.setVolume(1.0f);
					source.setPosition(entity.getPosition());
					source.play(breakingWood);
					source.setLoop(false);
				}
				break;
			default:
				break;
			}

			return true;
		} else {
			return false;
		}
	}

	/**
	 * Reproduces the sound of the entities
	 * 
	 * @param library
	 *            Set of sounds that can use to provide sound feedback to the
	 *            user
	 * @param entities
	 *            Set of entities that should provide sound feedback
	 * @param player
	 *            The player of the scene and also the listener
	 */
	private void playEntities(HashMap<AudioEnum, IAudioDescription> library, Set<Entity> entities, Player player) {
		if (!Utils.isEmpty(entities)) {
			Vector3f pPosition = player.getPosition();

			for (Entity entity : entities) {
				// Check is has one source
				IAudioSource source = sourcesAssign.get(entity);
				if (source == null) {
					continue;
				}

				if (!playCrash(library, entity, pPosition)) {
					// If not played a crash
					if (entity.getGenericEntity().getObjectType() != ModelEnum.tree) {
						continue;
					}

					// Make birds to play in each tree
					IAudioDescription breakingWood = library.get(AudioEnum.breakingWood);
					if (source.getAudioDescription() == breakingWood) {
						continue;
					}

					if (!source.isPlaying()) {
						IAudioDescription falconBuffer = library.get(AudioEnum.falcon);
						source.setPitch(1.0f);
						source.setVolume(1.0f);
						source.setPosition(entity.getPosition());
						source.play(falconBuffer);
						source.setLoop(true);
					}
				}

			}
		}
	}

	/**
	 * Plays all the sounds related with entities
	 * 
	 * @param library
	 *            Set of sounds that can use to provide sound feedback to the
	 *            user
	 * @param listener
	 *            Listener to put in a certain position
	 * @param entities
	 *            Set of entities that should provide sound feedback
	 * @param player
	 *            The player of the scene and also the listener
	 */
	public void play(HashMap<AudioEnum, IAudioDescription> library, IListener listener, Entity[] entities,
			Player player) {
		if (player == null) {
			return;
		}
		this.playPlayer(library, listener, player);
		this.assignSources(player, entities);
		this.playEntities(library, this.sourcesAssign.keySet(), player);

	}

	/**
	 * Clean up because we need to clean up when we finish the program
	 */
	public void dispose() {

	}
}
